### DISCUSSION TITLE
Any chance of improved developer tooling?

### DISCUSSION THREAD

**Discussion by mmarczell-graphisoft** [emscripten-core/emscripten#18215] [General]
The only way to debug WASM running in the browser is the closed source Chrome extension and that is quite limited and buggy. (Sometimes it doesn't even work, most of the time stepping doesn't work, standard library types are not inspectable etc.)

Would it be possible to write for example a VSCode extension for debugging WASM in the browser?
Would it be possible to write a memory graph inspector for C++ objects living in memory, similar to that of Xcode?

**Comment by Thaina** [comment]
I want to add that extension for intellisense in vscode would be great too, for writing macro in js file without error in vscode
